fix(detected_object_validation): apply transformation to shape.footprint.points

Shin-kyoto opened the pull request:
## Description

<!-- Write a brief description of this PR. -->
- Apply transformation(from object coordinate to map coordinate) to shape.footprint.points in object_lanelet_filter.
- If unknown object is not on lanelet, object_lanelet_filter removes it. However, sometimes unknown objects off the lanelet are not removed, as in the image below. 

![lanelet_filter_not_using_transform_footprint](https://user-images.githubusercontent.com/58775300/192789094-6575e8f3-94fb-4a34-90f2-69210e40b824.png)

- This is caused by `perception_utils::transformObjects`. This function transformed `object.kinematics.pose_with_covariance.pose` only,  and did not transform `object.shape.footprint.points`.
- When visualizing object in map coordinate, object rotated and intersected the lanelet. You can see object in map coordinate as a green LINE_LIST.

![image-20220928-061540](https://user-images.githubusercontent.com/58775300/192790511-abd80fa3-d0fa-4406-8b75-91f5e7f79a1c.png)

- So, in this PR,  I add process to transform `object.shape.footprint.points` and apply it in object_lanelet_filter.
- Unknown object not crossing the lanelet is removed by  object_lanelet_filter.

![lanelet_filter_using_transform_footprint](https://user-images.githubusercontent.com/58775300/192791127-08b36095-a949-4fe4-9503-1913ceccc83f.png)

## Related links
JIRA link: [TIERIV INTERNAL LINK](https://tier4.atlassian.net/browse/T4PB-21306)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

takayuki5168 commented:
@Shin-kyoto 
I guess `object.shape.footprint.points` is relative to the object pose. Therefore, if the frame of the object pose changes, we do not have to modify `object.shape.footprint.points`. Am I missing something?

Shin-kyoto commented:
> @Shin-kyoto I guess `object.shape.footprint.points` is relative to the object pose. Therefore, if the frame of the object pose changes, we do not have to modify `object.shape.footprint.points`. Am I missing something?

In my opinion, "`object.shape.footprint.points` is relative to the object pose." is right. But, if x-y axis rotates, relative coordinates (x, y) should also rotate.

![lanelet_filter_rotate](https://user-images.githubusercontent.com/58775300/192799068-e0f3e07a-24a0-4c86-876d-2a60f2fc7795.jpeg)

yukkysaito commented on common/perception_utils/include/perception_utils/perception_utils.hpp:
```diff
@@ -178,6 +180,25 @@ bool transformObjects(
       tf_target2objects = tf_target2objects_world * tf_objects_world2objects;
       tf2::toMsg(tf_target2objects, object.kinematics.pose_with_covariance.pose);
       // TODO(yukkysaito) transform covariance
+
+      if (transform_shape_footprint) {
+        // convert point from object coordinate(origin is
+        // object.kinematics.pose_with_covariance.pose) to target_frame_id
+        geometry_msgs::msg::TransformStamped tf_target2objects_stamped;
+        tf_target2objects_stamped.transform = tf2::toMsg(tf_target2objects);
+        for (auto & point : object.shape.footprint.points) {
+          // input_msg.header.frame_id to target_frame_id
+          geometry_msgs::msg::Point32 point_transformed;
+          tf2::doTransform<geometry_msgs::msg::Point32>(
+            point, point_transformed, tf_target2objects_stamped);
+
+          // target_frame_id to rotated object coordinate(origin is
+          // object.kinematics.pose_with_covariance.pose)
+          point.x = point_transformed.x - object.kinematics.pose_with_covariance.pose.position.x;
+          point.y = point_transformed.y - object.kinematics.pose_with_covariance.pose.position.y;
+          point.z = point_transformed.z - object.kinematics.pose_with_covariance.pose.position.z;
+        }
+      }
```
I have one question :pray: 
Why need this code? The shape footprint is based on object.kinematics.pose_with_covariance.pose as a specification of the message.

yukkysaito replied:
sorry, already commented by @takayuki5168 
https://github.com/autowarefoundation/autoware.universe/pull/1981#issuecomment-1260935186

yukkysaito replied:
@Shin-kyoto Can you talk about this later?

Shin-kyoto replied:
Removed this from perception_utils and added to detected_object_validation in  https://github.com/autowarefoundation/autoware.universe/pull/1981/commits/64d1c0d6ee4e057db6d6357d53b00a8cff591112

Shin-kyoto replied:
@yukkysaito
I used transformPoint and cared the quaternion in https://github.com/autowarefoundation/autoware.universe/commit/35fa236fb5c5c2e99368165d4adf01fabfa16bb7

Shin-kyoto commented:
## 10/05 updated
- Refactoring: Removed code to transform points from perception_utils and added to detected_object_validation in https://github.com/autowarefoundation/autoware.universe/commit/64d1c0d6ee4e057db6d6357d53b00a8cff591112
- Result: Unknown object not crossing the lanelet is removed by object_lanelet_filter.
![image](https://user-images.githubusercontent.com/58775300/194018692-6f8d33fa-316b-4f9e-94de-8dc7c296a9fa.png)

yukkysaito commented on perception/detected_object_validation/src/object_lanelet_filter.cpp:
```diff
@@ -23,6 +24,59 @@
+    const tf2_ros::Buffer &);
+  bool setTransformTarget2Objects(const geometry_msgs::msg::Pose &);
+  geometry_msgs::msg::Point32 transformPoint(const geometry_msgs::msg::Point32 &);
+};
+
+bool TransformFootprintPoint::setTransformTarget2ObjectsWorld(
+  const autoware_auto_perception_msgs::msg::DetectedObjects & input_msg,
+  const std::string & target_frame_id, const tf2_ros::Buffer & tf_buffer)
+{
+  const auto ros_target2objects_world =
+    getTransform(tf_buffer, input_msg.header.frame_id, target_frame_id, input_msg.header.stamp);
+  if (!ros_target2objects_world) {
+    return false;
+  }
+  tf2::fromMsg(*ros_target2objects_world, tf_target2objects_world_);
+
+  return true;
+}
+
+bool TransformFootprintPoint::setTransformTarget2Objects(const geometry_msgs::msg::Pose & pose)
+{
+  // get tf object to map
+  tf2::fromMsg(pose, tf_objects_world2objects_);
+  tf_target2objects_ = tf_target2objects_world_ * tf_objects_world2objects_;
+  tf_target2objects_stamped_.transform = tf2::toMsg(tf_target2objects_);
+
+  return true;
+}
+
+geometry_msgs::msg::Point32 TransformFootprintPoint::transformPoint(
+  const geometry_msgs::msg::Point32 & point)
+{
+  geometry_msgs::msg::Point32 point_transformed;
+  tf2::doTransform<geometry_msgs::msg::Point32>(
+    point, point_transformed, tf_target2objects_stamped_);
+
+  return point_transformed;
+}
+}  // namespace
+
```
I think no need to separate as the class. 
Can you use [this](https://github.com/autowarefoundation/autoware.universe/blob/main/common/perception_utils/include/perception_utils/perception_utils.hpp#L199-L228)?

yukkysaito commented on perception/detected_object_validation/src/object_lanelet_filter.cpp:
```diff
@@ -23,6 +24,59 @@
+    const tf2_ros::Buffer &);
+  bool setTransformTarget2Objects(const geometry_msgs::msg::Pose &);
+  geometry_msgs::msg::Point32 transformPoint(const geometry_msgs::msg::Point32 &);
+};
+
+bool TransformFootprintPoint::setTransformTarget2ObjectsWorld(
+  const autoware_auto_perception_msgs::msg::DetectedObjects & input_msg,
+  const std::string & target_frame_id, const tf2_ros::Buffer & tf_buffer)
+{
+  const auto ros_target2objects_world =
+    getTransform(tf_buffer, input_msg.header.frame_id, target_frame_id, input_msg.header.stamp);
+  if (!ros_target2objects_world) {
+    return false;
+  }
+  tf2::fromMsg(*ros_target2objects_world, tf_target2objects_world_);
+
+  return true;
+}
+
+bool TransformFootprintPoint::setTransformTarget2Objects(const geometry_msgs::msg::Pose & pose)
+{
+  // get tf object to map
+  tf2::fromMsg(pose, tf_objects_world2objects_);
+  tf_target2objects_ = tf_target2objects_world_ * tf_objects_world2objects_;
+  tf_target2objects_stamped_.transform = tf2::toMsg(tf_target2objects_);
+
+  return true;
+}
+
+geometry_msgs::msg::Point32 TransformFootprintPoint::transformPoint(
+  const geometry_msgs::msg::Point32 & point)
+{
+  geometry_msgs::msg::Point32 point_transformed;
+  tf2::doTransform<geometry_msgs::msg::Point32>(
+    point, point_transformed, tf_target2objects_stamped_);
+
+  return point_transformed;
+}
+}  // namespace
+
```
I think no need to separate as the class.
Can you use [this](https://github.com/autowarefoundation/autoware.universe/blob/main/common/perception_utils/include/perception_utils/perception_utils.hpp#L199-L228)?

Shin-kyoto replied:
It is difficult to use this.
That is because
- 1. I want to use `tf_target2objects_`, but [`transformObjects`](https://github.com/autowarefoundation/autoware.universe/blob/ce93396d7b17dbb37666a7251d148fa1dc26aa4d/common/perception_utils/include/perception_utils/perception_utils.hpp#L199-L228) does not return tf.
- 2. I want to use transformFootprintPoint in for-loop(objects) and for-loop(points). However, [`transformObjects`](https://github.com/autowarefoundation/autoware.universe/blob/ce93396d7b17dbb37666a7251d148fa1dc26aa4d/common/perception_utils/include/perception_utils/perception_utils.hpp#L199-L228) includes for-loop(objects). If I call [`transformObjects`](https://github.com/autowarefoundation/autoware.universe/blob/ce93396d7b17dbb37666a7251d148fa1dc26aa4d/common/perception_utils/include/perception_utils/perception_utils.hpp#L199-L228) in for-loop(points), it is not efficient.

Shin-kyoto replied:
@yukkysaito
I used transformPoint and cared the quaternion in https://github.com/autowarefoundation/autoware.universe/commit/35fa236fb5c5c2e99368165d4adf01fabfa16bb7

yukkysaito commented on perception/detected_object_validation/src/object_lanelet_filter.cpp:
```diff
@@ -102,8 +161,15 @@ void ObjectLaneletFilterNode::objectCallback(
+        input_msg->objects.at(index).kinematics.pose_with_covariance.pose);
+
       for (const auto & point : footprint.points) {
-        polygon.outer().emplace_back(point.x + position.x, point.y + position.y);
```
I think the problem is caused this code.
This code don't care the quaternion.

Shin-kyoto replied:
Yes. So I use tf to transform point with translation and rotational transfer.

Shin-kyoto replied:
@yukkysaito 
Use transformPoint and  care the quaternion in https://github.com/autowarefoundation/autoware.universe/pull/1981/commits/35fa236fb5c5c2e99368165d4adf01fabfa16bb7

yukkysaito commented on perception/detected_object_validation/src/object_lanelet_filter.cpp:
```diff
@@ -23,6 +24,22 @@
 
+namespace
+{
+geometry_msgs::msg::Point32 transformFootprintPoint(
+  const geometry_msgs::msg::Point32 & point, const geometry_msgs::msg::Pose & pose)
+{
+  tf2::Transform tf_data_;
+  geometry_msgs::msg::TransformStamped tf_data_stamped_;
+  tf2::convert(pose, tf_data_);
+  tf_data_stamped_.transform = tf2::toMsg(tf_data_);
+  geometry_msgs::msg::Point32 point_transformed;
+  tf2::doTransform<geometry_msgs::msg::Point32>(point, point_transformed, tf_data_stamped_);
+
+  return point_transformed;
+}
```
sample code
```suggestion
geometry_msgs::msg::Point32 transformPoint(const tf2::Transform & transform
  const geometry_msgs::msg::Point32 & point)
{
  tf2::Transform tf_point, tf_transformed_point;
  tf_point.setOrigin(point.x, point.y, point.z);
  tf_transformed_point = transform * tf_point;
  return convertToGeometrtMsg(tf_transformed_point)  
}
```

Shin-kyoto commented on perception/detected_object_validation/src/object_lanelet_filter.cpp:
```diff
@@ -23,6 +24,22 @@
+geometry_msgs::msg::Point32 transformFootprintPoint(
+  const geometry_msgs::msg::Point32 & point, const geometry_msgs::msg::Pose & pose)
+{
+  tf2::Transform tf_data_;
```
@yukkysaito
Thank you for your advise!
I used transformPoint in https://github.com/autowarefoundation/autoware.universe/commit/35fa236fb5c5c2e99368165d4adf01fabfa16bb7, so that code you advised is outdated.

yukkysaito reviewed (approved):
LGTM

Shin-kyoto commented:
@TakaHoribe @kenji-miyake 
I have to get code owner review, so can I get your review? I'm sorry for bothering you during your busy time. Thank you for your cooperation!

kenji-miyake reviewed (approved):
But it's okay for me to merge this as it is.

Shin-kyoto commented on common/tier4_autoware_utils/include/tier4_autoware_utils/geometry/geometry.hpp:
```diff
@@ -490,6 +491,18 @@ inline geometry_msgs::msg::Point transformPoint(
+    .x(transformed_point.x)
+    .y(transformed_point.y)
+    .z(transformed_point.z);
+}
```
@kenji-miyake 
Thank you for your advice! I made code simple in https://github.com/autowarefoundation/autoware.universe/pull/1981/commits/736e13f9e167b6135b58cc0b301ba277ee502c6a
